src/character.cpp

```diff
@@ -12830,3 +12832,274 @@ int Character::book_fun_for( const item &book, const Character &p ) const
+    return vision_mod <= 4;
+}
+
+} // end unnamed namespace
+
+using fail_reason = read_fail_reason;
+const type item_is_book                 { _item_is_book,                 fail_reason::item_not_readable };
+const type not_driving                  { _not_driving,                  fail_reason::driving };
+const type has_enough_morale            { _has_enough_morale,            fail_reason::not_enough_morale };
+const type has_enough_skill             { _has_enough_skill,             fail_reason::not_enough_skill };
+const type can_learn                    { _can_learn,                    fail_reason::cant_learn };
+const type doesnt_need_reading_glasses  { _doesnt_need_reading_glasses,  fail_reason::needs_reading_glasses };
+const type not_illiterate               { _not_illiterate,               fail_reason::illiterate };
+const type not_blind                    { _not_blind,                    fail_reason::blind };
+const type can_see_listener             { _can_see_listener,             fail_reason::cant_see_listener };
+const type not_too_dark                 { _not_too_dark,                 fail_reason::too_dark };
+
+bool can_be_assisted( read_fail_reason reason )
+{
+    using r = read_fail_reason;
+    switch( reason ) {
+        case r::illiterate:
+        case r::needs_reading_glasses:
+        case r::blind:
+            return true;
+
+        case r::item_not_readable:
+        case r::driving:
+        case r::not_enough_morale:
+        case r::not_enough_skill:
+        case r::cant_learn:
+        case r::cant_see_listener:
+        case r::too_dark:
+            return false;
+
+        default:
+            debugmsg( "Unhandled read_fail_reason" );
+            return false;
+    }
+}
+
+std::string get_fail_message( const read_criteria_context &ctx, read_fail_reason fail_reason )
+{
+    using r = read_fail_reason;
+
+    const Character &reader = ctx.reader;
+    const Character &listener = ctx.listener;
+    const item &book = ctx.reading_material;
+
+    const bool reading_to_self = ( reader.getID() == listener.getID() );
+
+    // the cases handled below are the things actually being used within the game
+    // since different contexts only cares about a subset of all fail reasons
+
+    if( reader.is_avatar() ) {
+        // does not matter if reading to self or not
+        switch( fail_reason ) {
+            case r::item_not_readable:
+                return string_format( _( "Your %s is not good reading material." ), book.tname() );
+            case r::driving:
+                return _( "It's a bad idea to read while driving!" );
+            case r::not_enough_morale:
+                return _( "What's the point of studying?  (Your morale is too low!)" );
+            case r::not_enough_skill: {
+                const auto &type = book.type->book;
+                const skill_id &skill = type->skill;
+                const std::string fmt = _( "%s %d needed to understand.  You have %d" );
```

Comment:
I think it would be better to clarify more on this sentence, for example `%s (level %d) needed to understand.`, so translators can understand what this sentence is talking about. Also it would be better to use positional placeholders like `%1$s (level %2$d) needed ... %3$d` so it allows translators to rearrange the order to fit in their languages, for example the parameters may be in the order of `需要%2$d級%1$s⋯` in some other language.

Reply:
I see. Didn't know there was such a feature.